Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -194,75 +324,53 @@ extension SCCoAPUDPTransportLayer: SCCoAPTransportLayerProtocol {
     }
 
 
-    public func sendCoAPData(_ data: Data, toEndpoint endpoint: NWEndpoint) throws {
-        let connection = mustGetConnection(forEndpoint: endpoint)
-        connection.send(content: data, completion: .contentProcessed{ [weak self] error in
-            guard let self = self else { return }
-            if error != nil {
-                self.transportLayerDelegate?.transportLayerObject(self, didFailWithError: error! as NSError)
+    /// Retrieves new message id for endpoint
+    /// There could be multiple clients using the same underlying transport, so it's import to have centralized
+    /// message ids issuance
+    public func getMessageId(for endpoint: NWEndpoint) -> UInt16 {
+        operationsQueue.sync { [unowned self] in
+            if let currentMessageId = self.messageIdsPerEndpont[endpoint] {
+                let newMessageId = (currentMessageId % 0xFFFF) + 1
+                self.messageIdsPerEndpont[endpoint] = newMessageId
+                return newMessageId
+            } else {
+                let newMessageId = UInt16(arc4random_uniform(0xFFFF))
+                self.messageIdsPerEndpont[endpoint] = newMessageId
+                return newMessageId
             }
-        })
+        }
     }
     
-    public func closeTransmission() {
-        operationsQueue.async(flags: .barrier) {[weak self] in
-            self?.listener?.cancel()
-            self?.listener = nil
-            self?.connections.forEach{
-                $0.value.cancel()
+    public func sendCoAPMessage(_ message: SCMessage, toEndpoint endpoint: NWEndpoint, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws {
```

Comment:
I still look at this logic as more suitable for the client than a transport but ok.